akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowMapAsyncSpec.scala

```diff
@@ -101,25 +97,54 @@ class FlowMapAsyncSpec extends StreamSpec {
       latch.countDown()
     }
 
-    "signal future failure asap" in assertAllStagesStopped {
-      val latch = TestLatch(1)
-      val done = Source(1 to 5)
-        .map { n ⇒
-          if (n == 1) n
-          else {
-            // slow upstream should not block the error
-            Await.ready(latch, 10.seconds)
-            n
-          }
+    "a failure mid-stream MUST cause a failure ASAP (stopping strategy)" in assertAllStagesStopped {
+      import system.dispatcher
+
+      // run it multiple times, it is inherently racy as it depends on order of futures completing
```

Comment:
Couldn't you make it deterministic by using promises, with their futures in the source, mapAsync with identity and then succeeding/failing the promises?